sklearn/preprocessing/tests/test_data.py

```diff
@@ -1952,6 +1955,98 @@ def test_one_hot_encoder_unknown_transform():
     assert_raises(ValueError, oh.transform, y)
 
 
+def check_categorical(X):
+    enc = CategoricalEncoder()
+    Xtr1 = enc.fit_transform(X)
+
+    enc = CategoricalEncoder()
+    Xtr2 = enc.fit_transform(X)
+
+    enc = CategoricalEncoder(sparse=False)
+    Xtr3 = enc.fit_transform(X)
+
+    assert_allclose(Xtr1.toarray(), Xtr2.toarray())
+    assert_allclose(Xtr1.toarray(), Xtr3)
+
+    assert sparse.issparse(Xtr1)
+    assert sparse.issparse(Xtr2)
+    return Xtr1.toarray()
+
+
+def test_categorical_encoder():
+    X = [['abc', 1, 55], ['def', 2, 55]]
+
+    Xtr = check_categorical(np.array(X)[:, [0]])
+    assert_allclose(Xtr, [[1, 0], [0, 1]])
+
+    Xtr = check_categorical(np.array(X)[:, [0, 1]])
+    assert_allclose(Xtr, [[1, 0, 1, 0], [0, 1, 0, 1]])
+
+    Xtr = CategoricalEncoder().fit_transform(X)
+    assert_allclose(Xtr.toarray(), [[1, 0, 1, 0,  1], [0, 1, 0, 1, 1]])
+
+
+def test_categorical_encoder_errors():
+
+    enc = CategoricalEncoder()
+    X = [[1, 2, 3], [4, 5, 6]]
+    enc.fit(X)
+
+    X[0][0] = -1
+    msg = re.escape('Unknown feature(s) [-1] in column 0')
+    assert_raises_regex(ValueError, msg, enc.transform, X)
+
+    enc = CategoricalEncoder(handle_unknown='ignore')
+    X = [[1, 2, 3], [4, 5, 6]]
+    enc.fit(X)
+    X[0][0] = -1
+    Xtr = enc.transform(X)
+    assert_allclose(Xtr.toarray(), [[0, 0, 1, 0, 1, 0], [0, 1, 0, 1, 0, 1]])
+
+
+def test_categorical_encoder_specified_categories():
+    X = np.array([['a', 'b']], dtype=object).T
+
+    enc = CategoricalEncoder(categories=[['a', 'b', 'c']])
+    exp = np.array([[1., 0., 0.],
+                    [0., 1., 0.]])
+    assert_array_equal(enc.fit_transform(X).toarray(), exp)
+
+    # don't follow order of passed categories, but sort them
+    enc = CategoricalEncoder(categories=[['c', 'b', 'a']])
+    exp = np.array([[1., 0., 0.],
+                    [0., 1., 0.]])
+    assert_array_equal(enc.fit_transform(X).toarray(), exp)
+
+    # multiple columns
+    X = np.array([['a', 'b'], ['A', 'C']], dtype=object).T
+    enc = CategoricalEncoder(categories=[['a', 'b', 'c'], ['A', 'B', 'C']])
+    exp = np.array([[1., 0., 0., 1., 0., 0.],
+                    [0., 1., 0., 0., 0., 1.]])
+    assert_array_equal(enc.fit_transform(X).toarray(), exp)
+
+    # when specifying categories manually, unknown categories should already
+    # raise when fitting
+    X = np.array([['a', 'b', 'c']]).T
+    enc = CategoricalEncoder(categories=[['a', 'b']])
+    assert_raises(ValueError, enc.fit, X)
+    enc = CategoricalEncoder(categories=[['a', 'b']], handle_unknown='ignore')
+    enc.fit(X)
```

Comment:
Currently this gives a row of all zeros (with indeed only two columns). Is this the desired outcome? (not sure what to do otherwise)